### ISSUE TITLE
[temp.deduct.type] p6 type name includes non-deduced context

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#4878]
[temp.deduct.type] p6 says
> When a type name is specified in a way that **includes a non-deduced context**, all of the types that comprise that type name are also non-deduced. However, a compound type can include both deduced and non-deduced types.

Consider this example
````cpp
template<class  U ,class T>
struct A{
    A(T){}
};
template<class U>
struct unknow_content{
    using type = U;
};

template<class U, class T>
void fun(A<U, typename unknow_content<T>::type>, T,U){}

int main(){
   fun(0,0,0);  // ill-formed
}
````
the type name `A<U, typename unknow_content<T>::type>` includes a non-deduced context `typename unknow_content<T>::type`, hence the `U` that comprises the type name `A<U, typename unknow_content<T>::type>` is also non-deduced?

**Comment by jwakely** [comment]
It can't be deduced from the first argument, but it can still be deduced from the third argument.

**Comment by jensmaurer** [comment]
Yes, the U in the first parameter is a non-deduced context, but it's deduced from the third parameter.

**Comment by xmh0511** [comment]
@jensmaurer  @jwakely  However, the [implementations](https://godbolt.org/z/Wdso85Pxr) say `U` cannot be deduced

> mismatched types 'A<U, typename unknow_content<T>::type>' and 'int'

**Comment by jensmaurer** [comment]
It seems "When a type name is specified in a way that includes a non-deduced context...", the "includes" is hazy here. Maybe "nearest enclosing type name" is what compilers implement.

**Comment by xmh0511** [comment]
The intent seems 
> When a type whose nested-name-specifier is a non-deduced context, all of the types that comprise that type name are also non-deduced. However, a compound type can include both deduced and non-deduced types.

The formal example indicates that.  
> [Example 2: If a type is specified as A<T>​::​B<T2>, both T and T2 are non-deduced. Likewise, if a type is specified as A<I+J>​::​X<T>, I, J, and T are non-deduced. If a type is specified as void f(typename A<T>​::​B, A<T>), the T in A<T>​::​B is non-deduced but the T in A<T> is deduced. — end example]  

Furthermore, `decltype(T{})::B<T2>`.
